### ISSUE TITLE
[Core] Support secrets in ray job submission requests

### ISSUE THREAD

**Issue by jhasm-ck** [ray-project/ray#39502]
### Description

Currently there is no way of passing secrets to a job submission request. This [doc](https://docs.ray.io/en/latest/ray-core/runtime_env_auth.html#running-on-kuberay-secrets-with-netrc) suggests every input passed in the request is visible on the dashboard and logs. Only secure way to use secrets in ray jobs is to mount them on the pods (when using kuberay), but this requires changing the mounts for every job.

We do have a headers input in the [JobSubmissionClient](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/doc/ray.job_submission.JobSubmissionClient.html#ray.job_submission.JobSubmissionClient) but these headers or cookies are not available to the job.

I suggest either making the headers and cookies available to the job or supporting a `secretes` input to the job submission client or method.

### Use case

1. Passing secrets to the ray jobs at the time of submitting jobs, instead of at the time of creating a cluster.
2. Not having to mount secrets on the ray cluster pods and accessing from a more secure secrets manager outside of ray.
3. Sharing the same ray cluster or cluster config between jobs and users, by changing entrypoint, inputs and secrets.
4. Building multi-tenant infrastructure of ray clusters, where secrets do not belong to a ray cluster, but rather belong to a job.

**Comment by jhasm-ck** [comment]
cc @richardliaw

**Comment by 0x2b3bfa0** [comment]
Duplicate of https://github.com/ray-project/ray/issues/28276?

**Comment by 0x2b3bfa0** [comment]
Related to https://github.com/ray-project/ray/issues/23150

**Comment by rynewang** [comment]
A `secret_env_vars` in runtime environment looks like the best way to do that: it's a natural extension to what we already offer (`env_vars`) and it's simple.

Behavior:
- part of the runtime env
- applied as environment variables just like `env_vars`
- but never shown in Ray logs or Dashboards.

One can specify them via the job submission client, or via the CLI `ray job submit --runtime-env`.

## Overview 

On client side it's just a plain, unparsed json. It's sent here and there until it reaches https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/agent/runtime_env_agent.py . 

`class RuntimeEnvAgent` is an HTTP server that handles creation and deletion of a runtime env. It's a pluggable structure: each kind of runtime env (e.g. `working_dir` or `pip`) is registered as a [`RuntimeEnvPlugin`](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/plugin.py) in the Agent's `__init__` constructor. The Agent forwards creation, deletion to the plugins.

On a new runtime env creation: the agent makes a `RuntimeEnvContext` which records what to do to create a worker process. Then it invokes each Plugin to prepare the files and prepend commands and env vars. Then the worker is started with all those things.

On a runtime env deletion: when all workers died, the Agent invokes each Plugin to delete the prepared files.


## The Plugin


We can create a new SecretEnvVars plugin like this:

- `class SecretEnvVars(RuntimeEnvPlugin):`
- `validate`: verifies we have a Dict from str to str.
- `get_uris`: return an empty array `[]`.
- `create`: no-op, return 0
- `modify_context`: extend the context.env_vars with the secret KVs.
- `delete_uri`: no-op, return 0

The new plugin can be a new file called `python/ray/_private/runtime_env/secret_env_vars.py`. Then we can ship it as a first party plugin, by adding it in the `RuntimeEnvAgent.__init__ function](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/agent/runtime_env_agent.py#L199) as a field and registering it there.

You can use the [working_dir](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/working_dir.py#L115) for reference. Note you don't have a "EnvVarPlugin" because we simply embed all env vars to the context in the first place.

## Logging

I think we have a bunch of logs logging the whole runtime env like [here](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/agent/runtime_env_agent.py#L354) (search "logger"). 

Short term: we hard code by never logging serialized runtime_env && only log the dict runtime_env with all entries whose keys contains "secret" removed.

Longer term (but still doable in one PR, your call): we add a method to `RuntimeEnvPlugin` namely `def show(runtime_env) -> str`. It defaults to return `json.dumps({self.name: runtime_env[self.name]})`, but for our plugin, it returns `json.dumps{"secret_env_vars": f"**{len(runtime_env["secret_env_vars"])} secrets**"}`

Next steps: write a unit test to load some secret env vars, and then read the logs to make sure all accidental logs are removed.

**Comment by jhasm-ck** [comment]
Thanks @rynewang, appreciate the details. I will go through them and come back with questions or a PR.

**Comment by jhasm-ck** [comment]
@rynewang FYI, for logging safety of secret_env_vars, I am thinking we store the values in `RuntimeEnv` dict with base64 encoding. That way we can be sure that we do not print them in plain text unintentionally in the logs or anywhere else. Since the `SecretEnvVarsPlugin` has control over how these secrets are set in the context, it can decode the values before setting. This way the secrets are encoded all the time, until the last moment.

This does not guarantee that the some code in worker does not print/log the environment variables, but that needs to be handled separately, IMO.

Let me know if you think otherwise.

**Comment by jhasm-ck** [comment]
One question on your suggestion @rynewang 
> Longer term (but still doable in one PR, your call): we add a method to RuntimeEnvPlugin namely def show(runtime_env) -> str. It defaults to return json.dumps({self.name: runtime_env[self.name]}), but for our plugin, it returns json.dumps{"secret_env_vars": f"**{len(runtime_env["secret_env_vars"])} secrets**"}

Th plugin just reads the secret_env_vars from [RuntimeEnv](https://github.com/ray-project/ray/blob/master/python/ray/runtime_env/runtime_env.py#L133) object (dict). So the secrets are stored in the [RuntimeEnv](https://github.com/ray-project/ray/blob/master/python/ray/runtime_env/runtime_env.py#L133) not in the plugin, right? In this case adding a `show()` method on the plugin will not help.

Out problem is code printing entire [RuntimeEnv](https://github.com/ray-project/ray/blob/master/python/ray/runtime_env/runtime_env.py#L133) dict which contains secrets as one entry. And in this class I see methods like `serialize`, `deserialize`, `to_dict()` which make me think it is pretty hard to avoid printing/leaking one entry from this dict.

I think storing the secrets encrypted (instead of base64 encoded) is a way of guaranteeing no leak, but it does add some complexity of storing the encryption key, which is why I am hesitant to take that route.

**Comment by pcmoritz** [comment]
One way how this could be done is have a datatype like
```
class ProtectedString:
    def __init__(self, s: Union[str, "ProtectedString"]):
        if isinstance(s, ProtectedString):
            self._UNSAFE_DO_NOT_USE: str = s._UNSAFE_DO_NOT_USE
        elif isinstance(s, str):
            self._UNSAFE_DO_NOT_USE = s
        else:
            raise ValueError(
                f"{type(s)}:'{repr(s)}' is not a valid type for ProtectedString"
            )

    def __eq__(self, other: Any) -> bool:
        if isinstance(other, ProtectedString):
            return self._UNSAFE_DO_NOT_USE == other._UNSAFE_DO_NOT_USE
        return False

    def __str__(self) -> str:
        error_msg = "ProtectedStrings should not be serialized directly."
        raise RuntimeError(error_msg)


    def __repr__(self) -> str:
        return str(self)
```
and use that for the secret values of those environment variables,  and similar in C++ code to make it less likely for the secrets to become part of log files (also avoid future Ray versions printing them by accident due to programmer errors).

This could be good enough for a good amount of simple applications. If you are really serious, you should of course use a secret store and then give the machines in the cluster that need access the right permissions through an IAM like system and we should point that out in the documentation.

**Comment by jhasm-ck** [comment]
That feels like opening a can of worms though, it is likely going to break a lot of code as we are just raising an error.

**Comment by jhasm-ck** [comment]
Thinking more about this, I am inclined towards storing encrypted secrets in RuntimeEnv object. Here is my reasoning,

1. We can either allow logging the secrets or not allow. 
2. If we want to do the latter, we have to make sure we change all the current log instances and make sure it is not logged in future. This will require an approach like mentioned above where we raise an error when someone tries to log the `ProtectedString`. This requires a significant labor upfront to correct logs and runs a risk of unintentional failures for edge cases. 
3. If we store the secrets encrypted we do not care about logging them because the printed value is not useful, assuming the encryption key is safe. We can still implement measures to avoid logging secrets, but they do not have to be disruptive.
4. Going with encryption solution is a two-way-door as it does not require too much of upfront investment and still allows us to block secrets logging in future, if needed. It will also be easy to remove the encryption/decryption part as it will be used at one place only.

Being that said, we do need to think about where to store the encryption key, to be able to support the encryption based solution. One way is to assume the encryption key will be made available in an environment variable on the ray cluster, which in kuberay world translates to mounting the key on the pod. It does feel like a little step backwards because mounting secrets on the pod is what we are trying to avoid. But in this case, it is limited to only one secret and that secret does not need to change or expire, but can be changed as needed. The encryption/decryption is all done before the job starts running, so this allows us to manage key in other ways and even change the key frequently.

In the above proposal for storing the key, I am assuming the `RuntimeEnvAgent` http server runs on the ray cluster and not on the client submitting job. Please correct me if I am wrong.

Curious to hear your thoughts and suggestions.

**Comment by rynewang** [comment]
Some thoughts:

1. we will not implement an "end to end encryption" for those secrets in foreseeable future. Most if not all things are transferred in cleartext.
2. That said, some simple measure of hiding them from accidental prints are welcome. i.e. No waterproof, but foolproof.
3. base64 lies quite well in the foolproof realm: you don't easily leak it out cleartext, but a hacker can still manually read it out. (maybe we can do some little hardening/obfuscation on base64, e.g. like this

```

def encode_with_shift(data, shift=1):
    b64_encoded = base64.b64encode(data.encode())
    shifted = bytearray([b + shift for b in b64_encoded])
    return shifted

def decode_with_shift(data, shift=1):
    unshifted = bytearray([b - shift for b in data])
    return base64.b64decode(unshifted).decode()

```

4. after this PR I will investigate a little more on @pcmoritz 's idea: we can rewrite `__repr__` and `__str__` to something like `return f"<{len(d)} secrets>"` which can tell us if the bits are set without telling us the fields or the values. I need to think more about how this also works in client side, not the runtime env agent side. We can do both base64 and this print-hiding.
5. In raylet (C++ side), we print the runtime envs as a string in DEBUG mode but not in INFO mode. This should be fine (DEBUG mode == no security).

So overall the PR's good. I will approve the PR once it's in good shape.

**Comment by rynewang** [comment]
cc @thomasdesr

**Comment by jhasm-ck** [comment]
@rynewang Makes sense. Thanks for the clarity!
Also, we are going to remove the `secret_env_vars` from the ray dashboard, right? Otherwise the base64 will not be sufficient because it is to easy to manually see the encoded secrets. 

Copying from logs require a little more work and we are working on removing those as well, like you said, but showing on the dashboard is just too easy to grab.

**Comment by rynewang** [comment]
re Dashboard: yes, we will do that after this PR is out & we agree on a protocol for those secrets. cc @alanwguo

**Comment by 0x2b3bfa0** [comment]
Obfuscated Base64 encoding is absolutely no substitute for proper cryptography. I'd suggest to either use [SealedBox](https://pynacl.readthedocs.io/en/latest/public/#nacl-public-sealedbox)[^1] or just assume the risks of plain text and e.g. mask[^2] the sensitive values on logs.

On virtual machine Ray deployments, we could probably use SSH keys to encrypt/decrypt data. On Kuberay we'd have to figure out something different, though.

[^1]: See for instance [how GitHub uses `libsodium` sealed boxes for Actions Secrets](https://docs.github.com/en/rest/guides/encrypting-secrets-for-the-rest-api?apiVersion=2022-11-28).
[^2]: It would be nice to mask sensitive values on logs and similar information, regardless of the chosen approach. Note, though, that secret masking isn't precisely trivial; see e.g. [this GitHub Actions Runner test](https://github.com/actions/runner/blob/121f080023450ab0ee5fc2d7961d5b630434d41f/src/Test/L0/Worker/ActionCommandManagerL0.cs#L424-L444) for an example.

**Comment by thomasdesr** [comment]
Hey!

Just catching up but I'd prefer if we took a bit more time on this. Maybe start a REP for the design?

Secret stores are a pretty complex space and we should try to have some sort of intentional answer for a number of capabilities beyond confidentiality. E.g. Auditability (who accessed what, when) and Authorization (should X be allowed access to Y), neither of which are currently upon in this proposal.

**Comment by jhasm-ck** [comment]
Hey @thomasdesr just to be clear, we are not proposing a secret store here, just trying to build a pass-through plumbing in the ray jobs so users have a way to submit secrets to ray jobs. These secrets are not supposed to be stored anywhere, just used in the job and then discarded.

**Comment by thomasdesr** [comment]
Okay, finished reading & thinking. I think something like this is going to be pretty hard to build into Ray as a first class citizen.

As briefly mentioned before, Secret management presents a pretty long list of challenges they need to solve. If you're curious there was a nice talk at AppSec 2015 on this:
* [slides](http://schd.ws/hosted_files/appsecusa2015/a5/Turtles.pdf)
* [video](https://www.youtube.com/watch?v=OUSvv2maMYI)

Also, more Ray specific, Ray's current design is built around a single global code execution domain. It is extremely difficult to create effective boundaries inside a code execution domain (as evidenced by the amount of effort that has been spent trying to stop exploits in browsers and hypervisors). So, unfortunately, exposing secrets to a cluster exposes them to everything that has run on the cluster, and probably everything that will run on that cluster. Until we can add some form of effective internal isolation into a Ray, I would discourage exposing a "job-scoped secret" concept since it can't actually be achieved. We’ve discussed on and off for years what it would take but this likely isn’t a near-term change. xD

The solution I'd strongly encourage is to lean into whatever existing secret management tools you have at your disposal (if you don’t have one the cloud provider’s tools are decent); keeping in mind that the smallest unit of isolation you can possibly rely on is the cluster level.

If we were to build something and label it a "secret" system, I'd want to be able to have a path to doing a decent job at it and I don't see a path to that right now without structural changes to Ray. I don't want perfect to be the enemy of the good enough though so if anyone has any idea they like please share; but I am cautious about having an API that can’t do what it seems to plainly offer.

But if you're happy with the properties the e2e encrypted blob idea presents, you can totally do that today without code changes to Ray. With one caveat, I'd probably toss the encrypted blobs somewhere in RuntimeEnv; but be cautious about where you're storing the Runtime env (it defaults to in GCS volatile memory, but can also easily be somewhere semi-permanent like a cloud blob store).

That said, I’d still encourage you to use your existing answer to secrets management as you normally would (but again, keeping in mind Ray’s isolation boundary is the cluster lifetime).

**Comment by thomasdesr** [comment]
Thanks for that clarification :D

Even with that in mind, its still kinda a secret store because something within Ray has to hold onto these values for tasks or hosts that haven't started yet unless the secrets are externalized from Ray and held elsewhere.

**Comment by jhasm-ck** [comment]
How can I access the external secret manager from my ray job without the auth token? And I cannot mount the auth token on the pod as they expire. Only way I can achieve this is by creating a new ray cluster for every job, please correct me if I am wrong.

Using external secret manager is what I want to do, I wouldn't trust anything else to store secrets. I am just hoping to add plumbing to the ray jobs so users can pass the auth token to the job and then the job can access the secrets from the secret manager. So what I really mean by "secrets" in this change is just the auth token.

**Comment by thomasdesr** [comment]
Ahh, how to bootstrap initial access. Sharing general advice here: I'd punt the "how do I get the first secret" into your cloud provider because they've already solved that problem.

I'd like to get you unblocked but that probably involves chatting about the specifics of your setup. Do you mind shooting me an email to <thomas at anyscale.com> so we discuss specifics?

**Comment by jhasm-ck** [comment]
Sure, let’s discuss offline. On Oct 23, 2023, at 9:56 PM, Thomas Desrosiers ***@***.***> wrote:﻿
Ahh, how to bootstrap initial access. Sharing general advice here: I'd punt the "how do I get the first secret" into your cloud provider because they've already solved that problem.
I'd like to get you unblocked but that probably involves chatting about the specifics of your setup. Do you mind shooting me an email to  so we discuss specifics?

**Comment by jhasm-ck** [comment]
I have given some more thoughts on this, and here is a summary
1. From ray side this looks like a bigger project, viewed as supporting a secret store inside the ray cluster.
2. If we want to do a good job a protecting secrets we need to use encryption, not encoding.
3. The challenging part with encryption is the key management, and supporting a scheme that works for everyone is going to be involved work.
4. If I am going to use the encryption, I may not need to add `secret_env_vars`, I can just use the `env_vars` and pass encrypted values. That way the encryption and decryption can be handled outside of ray. The client (or a proxy service between the client and the ray cluster) needs to encrypt the secret vars and the ray job needs to be able to decrypt it. I agree this puts more work on the ray users, but the upside is that everyone can manage the encryption part according to what fits well with the rest of the infrastructure they have.

In my case, I already have a proxy service (api gateway) that sits between the ray cli/sdk and ray cluster (head service). This is a standard practice at my firm. Currently this already enforces authentication by requiring users to pass their auth token in the `--headers` input to the `ray job submit` call. This headers do not reach the ray job, which was the premise of my feature request, and I was hoping to add some plumbing to make this reach. In the first discussion with the ray team it looked like that headers is not intended to be used by the ray jobs so we should add a new runtime_env called `secret_env_var`. But now it sounds like I am coming to a full circle and going to rely on the headers, but without making any changes to ray code.

Here is my plan,
1. The proxy service will read the auth token from headers, encrypt it (using a public key) and inject it into the request at `runtime_env.env_vars.AUTH_TOKEN`. This request gets forwarded to the ray cluster (head service).
2. The ray job will call another service to decrypt the token. E.g. `token = ray_encryption_service.decrypt(os.environ['AUTH_TOKEN'])`. The job can then use this token to pull secrets from the secrets manager and it will have the same privileges as the owner of the auth token.
3. Since the ray infrastructure will never see the decrypted token there is no risk with printing in logs or showing it on the dashboard.
4. The value stored in environment variable on the ray nodes is encrypted as well.

NOTE: In this case requiring the user to pass the encrypted auth token in the env_var input will not work, if this ray cluster is shared between multiple users. This is because the encrypted value will show up on the ray dashboard in `runtime_env`, and another user can use that in their request to impersonate the first user. The encryption needs to happen in the proxy service to make sure the user has access to the raw auth token. And of course, we are assuming the plain text token is transmitted securely from the ray cli/sdk to the proxy service (e.g. using TLS).

**Comment by jhasm-ck** [comment]
As an orthogonal idea @thomasdesr what do you think about simply redacting JWT tokens from `runtime_env` when serializing in string or json form? 
Looks like I am doing everything just to avoid my JWT token in `env_var` being exposed to ray dashboard and logs. I am fine with keeping the token in memory. Encryption mechanisms add more complexity and printing/showing a JWT token somewhere is not really a feature. So I think just not showing those tokens will be a good default behavior for everyone and perfectly solves my problem.

Following is an example of how this might look like,
```python
import json
import re

def redact_jwt_tokens(payload):
    # Regular expression pattern to match JWT tokens
    jwt_pattern = r'[\w-]+\.[\w-]+\.[\w-]+'
    # Replace JWT tokens with a redacted string (e.g., "[REDACTED]")
    jwt_matches = re.findall(jwt_pattern, payload)
    redacted_payload = re.sub(jwt_pattern, '[REDACTED]', payload)

    return redacted_payload, jwt_matches

# Sample JSON payload containing a JWT
json_payload = '{"user": "john.doe", "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.5Tc5MnBsB1aIzZqZVu_UjNtVcGc7Y9AlteG3eJuXTOE"}'

# Detect and redact JWT tokens
redacted_payload, jwt_tokens = redact_jwt_tokens(json_payload)

# Print the redacted payload and extracted JWT tokens
print(f"Original payload: {json_payload}")
print(f"\nRedacted Payload: {redacted_payload}")
print(f"\nExtracted JWT Tokens: {jwt_tokens}")
```

If the regex is seems too wide and can cause unintended values to be redacted, we could add a feature flag to disable this behavior when needed.
